File: replay-vk/src/main/java/dev/luna5ama/glc2vk/replay/MemorySuballocator.java
package dev.luna5ama.glc2vk.replay;

public class MemorySuballocator {
    public MemorySuballocator(long baseAddress) {
        this.baseAddress = baseAddress;
    }

    private final long baseAddress;
    private long allocatedSize = 0;

    public long allocate(long byteSize, long byteAlignment) {
        byteAlignment = Math.max(byteAlignment, 8);
        long baseAddress = this.baseAddress;
        long sliceOffset = ((baseAddress + allocatedSize + byteAlignment - 1) & -byteAlignment) - baseAddress;
        allocatedSize = sliceOffset + byteSize;
        return sliceOffset;
    }

    public long getAllocatedSize() {
        return allocatedSize;
    }
}

File: replay-vk/src/main/kotlin/dev/luna5ama/vibris/replay/MemorySuballocator.kt
package dev.luna5ama.vibris.replay

class MemorySuballocator(private val baseAddress: Long) {
    var allocatedSize = 0L
        private set

    fun allocate(byteSize: Long, requestedAlignment: Long): Long {
        val alignment = maxOf(requestedAlignment, 8L)
        val sliceOffset = ((baseAddress + allocatedSize + alignment - 1L) and -alignment) - baseAddress
        allocatedSize = sliceOffset + byteSize
        return sliceOffset
    }
}